---
title: 'SciX: Genies'
author: "Lachlan Gray"
date: "19/06/2022"
output: html_document
runtime: shiny
---
# Introduction
The central dogma of biology states that DNA is transcribed into RNA which is tranlated into protein. RNA sequencing allows scientists to infer the proteins being produced within the tissues of an individual. By comparing the expression levels of RNA between individuals with disease and healthy controls we can identify potential causes of disease or therapeutic drug targets.

In this study, we have processed raw RNA sequencing data from the brains of people with neurodegenerative disease and healthy controls. These studies include:
* Parkinson's disease
* Alzheimer's disease
* Hungtington's disease

To process the raw RNA sequencing reads we have done the following:
1. Files were downloaded from NCBI GEO
2. Sequencing reads were trimmed with Trimmomatic to remove adaptors and low quality reads
3. Reads were mapped to the GRCh38 Human genome with STAR
4. Transcripts counted and annotated with Stringtie

The comma seperated file (.csv) we provide is the output from Stringtie which contains the raw counts of each RNA transcript sequenced in each sample. 

To run the following code, press the green arrow and R will run the code for you. All of the plots and 

```{r setup, include=FALSE}
install.packages('knitr', verbose=F)
knitr::opts_chunk$set(echo = TRUE)
r <- getOption("repos")
r["CRAN"] <- "https://cran.csiro.au/"
options(repos = r)
```

### Install packages
```{r, echo=FALSE}
install.packages('ggplot2')
install.packages('reshape2')
install.packages('ggrepel')
install.packages('ggupset')
install.packages('BiocManager')
install.packages('statmod')
install.packages('pheatmap')
BiocManager::install("qvalue")
BiocManager::install("edgeR")
BiocManager::install("clusterProfiler")
```

### Load the packages
```{r echo=FALSE}
library(ggplot2, verbose = F)
library(qvalue, verbose = F)
library(reshape2, verbose = F)
library(ggrepel, verbose = F)
library(ggupset, verbose = F)
library(enrichplot, verbose = F)
library(edgeR, verbose = F)
library(DOSE, verbose = F)
library(clusterProfiler, verbose = F)
library(statmod, verbose=F)
library(pheatmap, verbose=F)
```

## Set working directory
We need to tell R which folder (known as directory) our data is located and where we want to store our results.
This directory will be the unzipped SciX-main directory we just downloaded.  
As naming paths are different in Mac and PC please make sure you run the appropriate line of code.

### For Mac
```{r echo=FALSE}
setwd('~/Desktop/SciX-main/')
```

### For PC
```{r echo=FALSE}
setwd(paste0(file.path(Sys.getenv("USERPROFILE"),"Desktop"), 'SciX-main'))
```

### Read in RNA sequencing count matrix
```{r echo=FALSE}
count.matrix <- read.csv('HD.data.csv', 
                         header=T, row.names = 1)
```

### Read in sample metadata
```{r echo=FALSE}
metadata <- read.delim('HD.metadata.txt')
print(metadata)
```

### The count matrix has  gene names as rownames and  sample names as column names
```{r echo=FALSE}
count.matrix[1:5,]
```

### To make the column names more informative we replace with metadata$condition column
```{r echo=FALSE}
colnames(count.matrix) <- metadata$condition
# Print first 5 genes
count.matrix[1:5,]
```

### By looking at expression of XIST we can determine sex of samples
```{r echo=FALSE}
XIST.counts <- count.matrix[grep('XIST', rownames(count.matrix)),]
XIST.counts
metadata$sex <- apply(XIST.counts, 1, function(x) ifelse(x> 20, 'F', 'M'))
```

# Differential expression analysis with edgeR likelihood ratio test
We will perform a statistical test to determine which genes are different between our conditions. For this, we will use the likelihood ratio test which ...
Before we start we have to tell R the condition and sex of each sample.
We then make our disease samples the reference group. This tells us the difference in gene expression in relation to our disease group. For example, a gene with a poitive logFC is upregulated in disease and a negative logFC is downregulated in disease.
We then filter out lowly expressed genes, normalise the expression values and perform the test.
To visualise our results, we create plots to show differentially expressed genes.

```{r echo=FALSE}
targets = data.frame(group = gsub('[0-9]+', '', 
                    metadata$condition),
                    sex = metadata[,3])
colnames(targets) <- c("group", "sex")
design <- model.matrix(~0+sex+group, data=targets)
y = DGEList(counts = count.matrix, group = targets$group)
y$samples$group <- factor(y$samples$group, levels=c('H', 'C'))
keep <- filterByExpr(y)
y <- y[keep, ,keep.lib.sizes=FALSE]
y <- calcNormFactors(y, method = "TMM")
tmm <- cpm(y)
y <- estimateDisp(y, design, robust=TRUE)
fit <- glmQLFit(y, design)
lrt <- glmLRT(fit, coef=ncol(fit$design))
tt <- topTags(lrt, n=nrow(y), p.value=0.05)
# Smearplot
plotSmear(lrt, de.tags=rownames(tt$table))
pdf('smearplot.pdf')
plotSmear(lrt, de.tags=rownames(tt$table))
dev.off()
print(summary(decideTests(lrt)))
lrt <- as.data.frame(lrt)
lrt$FDR <- qvalue(lrt$PValue)$qvalues
gene <- gsub('ENSG[0-9]+.[0-9]+\\|', '', rownames(lrt))
lrt <- cbind(gene,lrt)
# Top 10 differentially expressed genes
lrt[order(lrt$FDR),][1:10,]

# MDS plot
mds <- plotMDS(y, top=1000, plot=F)
toplot <- data.frame(Dim1 = mds$x, Dim2=mds$y, group=factor(y$samples$group), sex=factor(metadata$sex))
ggplot(toplot, aes(Dim1, Dim2, colour=group)) + 
  geom_point(aes(shape=sex), size=5) +
  scale_shape_manual(values=c(17, 15)) +
  geom_text_repel(data=toplot, label=rownames(toplot)) +
  ggtitle('Multidimensional scaling (MDS) plot')

pdf('MDS.plot.pdf')
ggplot(toplot, aes(Dim1, Dim2, colour=group)) + 
  geom_point(aes(shape=sex), size=5) +
  scale_shape_manual(values=c(17, 15)) +
  geom_text_repel(data=toplot, label=rownames(toplot)) +
  ggtitle('Multidimensional scaling (MDS) plot')
dev.off()
```

### Save result file to working directory
```{r echo=TRUE}
write.table(lrt, row.names = F, sep = "\t", 'edgeR-LRT.HD.txt')
```

## Displaying results in volcano plot
This plot displays the log fold-change and false dicovery rate for each gene. You can select the number of genes to label with the **n.genes** variable below.
```{r echo=FALSE}
n.genes = 20

threshold <- lrt$FDR < 0.05 & abs(lrt$logFC) > 0.5
lrt$threshold <- threshold

# Order on FDR
ordered <- lrt[order(lrt$FDR),]
ggplot(lrt) +
  geom_point(aes(x=logFC, y=-log10(FDR), colour=threshold)) +
  geom_text_repel(data=head(ordered, n.genes), aes(x=logFC, y=-log10(FDR), label=gene)) +
  ggtitle("Volcano Plot: Huntington's Disease") +
  xlab("log2 fold change") +
  ylab("-log10 FDR") +
  theme(plot.title = element_text(size = rel(1.5), hjust = 0.5),
        axis.title = element_text(size = rel(1.25))) +
  scale_color_discrete(name = "DEG")

pdf('Volcano.Plot.pdf')
ggplot(lrt) +
  geom_point(aes(x=logFC, y=-log10(FDR), colour=threshold)) +
  geom_text_repel(data=head(ordered, n.genes), aes(x=logFC, y=-log10(FDR), label=gene)) +
  ggtitle("Volcano Plot: Huntington's Disease") +
  xlab("log2 fold change") +
  ylab("-log10 FDR") +
  theme(plot.title = element_text(size = rel(1.5), hjust = 0.5),
        axis.title = element_text(size = rel(1.25))) +
  scale_color_discrete(name = "DEG")
dev.off()
```

## Display results as heatmap
This plot will show gene expression in a heatmap where the colour of the tile indicates gene expression in each sample. You can select the number of genes to show with the **n.genes** variable below.

```{r echo=FALSE}
n.genes <- 20
mat <- tmm[head(order(lrt$FDR), n.genes),]
rownames(mat) <- gsub('ENSG[0-9]+.[0-9]+\\|', '', rownames(mat))
pheatmap::pheatmap(mat)

pdf('heatmap.pdf')
pheatmap::pheatmap(mat)
dev.off()
```

# Over Representation Analysis (ORA)

```{r echo=FALSE}
gene.set <- read.gmt('reactome.gmt')
sigOE <- subset(lrt, abs(logFC) > 0.5 & FDR < 0.05)

ego <- enricher(gene = sigOE$gene, 
         universe = lrt$gene,
         pAdjustMethod = "fdr", 
         qvalueCutoff = 0.01,
         TERM2GENE = gene.set)

dotplot(ego, showCategory=20, orderBy="GeneRatio")
pdf('reactome.ORA.pdf', width = 14)
dotplot(ego, showCategory=20, orderBy="GeneRatio")
dev.off()

upsetplot(ego)
pdf('reactome.upsetplot.pdf')
upsetplot(ego)
dev.off()
```

### Match genes to DisGeneNet and perform chi-squared test
```{r echo=FALSE}
disgene <- read.delim('disgenenet.HD.tsv', header=T, sep="\t")
disgene <- unique(disgene$Gene)

deg <- subset(lrt, abs(logFC) > 0.5 & FDR < 0.05)
not.deg <- subset(lrt, FDR > 0.05)
up <- subset(lrt, logFC > 0.5 & FDR < 0.05)
down <- subset(lrt, logFC < -0.5 & FDR < 0.05)

# calculate numbers for different conditions
a <- lrt[lrt$FDR < 0.05 & abs(lrt$logFC) > 0.5 & lrt$gene %in% disgene,]$gene
b <- lrt[lrt$FDR < 0.05 & abs(lrt$logFC) > 0.5 & !(lrt$gene %in% disgene),]$gene
c <- lrt[lrt$FDR > 0.05 & lrt$gene %in% disgene,]$gene
d <- lrt[lrt$FDR > 0.05 & !(lrt$gene %in% disgene),]$gene
ctable <- matrix(c(length(a),length(b),length(c),length(d)),nrow=2, ncol=2)
chisq <- chisq.test(ctable)

# Expected values
chisq$expected
# Observed values
chisq$observed
# Pearson residuals. Positive values indicate a positive association
chisq$residuals

# ORA of significant genes known to cause disease
ego <- enricher(gene = a, 
                universe = lrt$gene,
                pAdjustMethod = "fdr", 
                qvalueCutoff = 0.01,
                TERM2GENE = gene.set)
dotplot(ego, orderBy="GeneRatio", font.size=10, label_format=20)
pdf('disgenenet.ORA.pdf', width=14)
dotplot(ego, orderBy="GeneRatio", font.size=12, label_format=20)
dev.off()
```

### Extract genes from interesting pathway. Select pathway with **pathway** variable
```{r echo=FALSE}
pathway <- 1
plot.genes <- unlist(strsplit(data.frame(ego)[pathway,8], '/'))
gene.name <- 
rownames(tmm) <- gsub('ENSG[0-9]+.[0-9]+\\|', '', rownames(tmm))
colnames(tmm) <- gsub('[0-9]+', '', colnames(tmm))
plot.data <- subset(tmm, rownames(tmm) %in% plot.genes)
plot.data <- plot.data[order(rowSums(plot.data), decreasing = T),]
plot.data <- melt(plot.data)
colnames(plot.data) <- c('gene', 'condition', 'cpm')
ggplot(plot.data, aes(x=gene, y=cpm, fill=condition)) + 
  geom_boxplot() +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  ggtitle(paste('', ego[pathway,1])) +
  ggtitle(paste('Expression of', ego[pathway,1], 'genes'))

pdf('boxplot.pdf', width=14)
ggplot(plot.data, aes(x=gene, y=cpm, fill=condition)) + 
  geom_boxplot() +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  ggtitle(paste('', ego[pathway,1])) +
  ggtitle(paste('Expression of', ego[pathway,1], 'genes'))
dev.off()

# Show logFC of genes in pathway
subset(lrt, gene %in% plot.data$gene)
```

